src/backend/vulkan/src/mirror.rs

```diff
@@ -0,0 +1,386 @@
+    variables: Vec<Variable>,
+    types: Vec<Type>,
+}
+
+pub fn reflect_spirv_module(code: &[u8]) -> SpirvReflection {
+    use spirv_utils::instruction::Instruction;
+
+    let module = spirv_utils::RawModule::read_module(code).unwrap();
+
+    let mut entry_points = Vec::new();
+    let mut variables = Vec::new();
+    let mut types = Vec::new();
+    for instr in module.instructions() {
+        match *instr {
+            Instruction::EntryPoint { execution_model, ref name, ref interface, .. } => {
+                if let Some(stage) = map_execution_model_to_stage(execution_model) {
+                    entry_points.push(EntryPoint {
+                        name: name.clone(),
+                        stage: stage,
+                        interface: interface.clone(),
+                    });
+                }
+            },
+            Instruction::Variable { result_type, result_id, storage_class, .. } => {
+                let decoration = map_decorations_by_id(&module, result_id.into());
+                let ty = {
+                    // remove indirection layer as the type of every variable is a OpTypePointer
+                    let ptr_ty = module.def::<desc::TypeId>(result_type.into()).unwrap();
+                    match *ptr_ty {
+                        Instruction::TypePointer { ref pointee, .. } => *pointee,
+                        _ => unreachable!(), // SPIR-V module would be invalid
+                    }
+                };
+
+                // Every variable MUST have an name annotation
+                // `glslang` seems to emit empty strings for uniforms with struct type,
+                // therefore we need to retrieve the name from the struct decl
+                let name = {
+                    let name = map_name_by_id(&module, result_id.into()).unwrap();
+                    if name.is_empty() {
+                        map_name_by_id(&module, ty.into()).unwrap()
+                    } else {
+                        name
+                    }
+                };
+
+                variables.push(Variable {
+                    id: result_id.into(),
+                    name: name.into(),
+                    ty: ty.into(),
+                    storage_class: storage_class,
+                    decoration: decoration,
+                });
+            },
+
+            _ => {
+                // Reflect types
+                if let Some(ty) = map_instruction_to_type(&module, instr) {
+                    types.push(ty);
+                }
+            },
+        }
+    }
+
+    SpirvReflection {
+        entry_points: entry_points,
+        variables: variables,
+        types: types,
+    }
+}
+
+pub fn populate_info(info: &mut shade::ProgramInfo, stage: shade::Stage, reflection: &SpirvReflection) {
+    if stage == shade::Stage::Vertex {
+        // record vertex attributes
+        if let Some(entry_point) = reflection.entry_points.iter().find(|ep| ep.name == "main" && ep.stage == stage) {
```

Comment:
Sounds like it should be using 'expect()' then.

> On Sep 5, 2016, at 17:35, msiglreith <notifications@github.com> wrote:
> 
> In src/backend/vulkan/src/mirror.rs:
> 
> > +                }
> > +            },
> > +        }
> > +    }
> > +
> > +    SpirvReflection {
> > +        entry_points: entry_points,
> > +        variables: variables,
> > +        types: types,
> > +    }
> > +}
> > +
> > +pub fn populate_info(info: &mut shade::ProgramInfo, stage: shade::Stage, reflection: &SpirvReflection) {
> > +    if stage == shade::Stage::Vertex {
> > +        // record vertex attributes
> > +        if let Some(entry_point) = reflection.entry_points.iter().find(|ep| ep.name == "main" && ep.stage == stage) {
> Assuming our module was compiled via glslang it shouldn't be possible. SPIR-V allows to specify any name for entrypoints (not only main). If there is no main() the code will probably fail somewhere else I assume during the pipeline creation as we are also asking for the main function there.
> https://github.com/gfx-rs/gfx/blob/master/src/backend/vulkan/src/factory.rs#L241
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>